File: cloudpilot-ionic/src/app/helper/Fifo.ts
export class Fifo<T> {
    constructor(size: number) {
        this.buffer = new Array(size);
    }

    push(value: T): this {
        if (this.size < this.buffer.length) {
            this.buffer[(this.start + this.size++) % this.buffer.length] = value;
        } else {
            this.start = (this.start + 1) % this.buffer.length;

            this.buffer[(this.start + this.size - 1) % this.buffer.length] = value;
        }

        return this;
    }

    pop(): T | undefined {
        if (this.size === 0) return undefined;

        const elt = this.buffer[this.start];

        this.start = (this.start + 1) % this.buffer.length;
        this.size--;

        return elt;
    }

    peekLast(): T | undefined {
        if (this.size === 0) return undefined;

        return this.buffer[(this.start + this.size - 1) % this.buffer.length];
    }

    replaceLast(value: T): this {
        if (this.size > 0) {
            this.buffer[(this.start + this.size - 1) % this.buffer.length] = value;
        }

        return this;
    }

    count(): number {
        return this.size;
    }

    flush(): this {
        this.start = 0;
        this.size = 0;

        return this;
    }

    private buffer: Array<T>;

    private start = 0;
    private size = 0;
}

File: src/emulator/Fifo.cpp
#include "Fifo.h"

Fifo::Fifo(uint32 capacity) : capacity(capacity) {
    buffer = new uint32[capacity];
    memset(buffer, 0, sizeof(uint32) * capacity);
}

Fifo::~Fifo() { delete[] buffer; }

void Fifo::Push(uint32 value) {
    buffer[(start + size) % capacity] = value;

    if (size < capacity)
        size++;
    else
        start = (start + 1) % capacity;
}

uint32 Fifo::Pop() {
    if (size == 0) return 0;

    const uint32 value = buffer[start];
    start = (start + 1) % capacity;

    size--;

    return value;
}

uint32 Fifo::Size() const { return size; }

void Fifo::Clear() {
    size = 0;
    start = 0;
}
